substrate/frame/bounties/src/lib.rs

```diff
@@ -1002,57 +1457,154 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
-	fn spend_funds(
-		budget_remaining: &mut BalanceOf<T, I>,
-		imbalance: &mut PositiveImbalanceOf<T, I>,
-		total_weight: &mut Weight,
-		missed_any: &mut bool,
-	) {
-		let bounties_len = BountyApprovals::<T, I>::mutate(|v| {
-			let bounties_approval_len = v.len() as u32;
-			v.retain(|&index| {
-				Bounties::<T, I>::mutate(index, |bounty| {
-					// Should always be true, but shouldn't panic if false or we're screwed.
-					if let Some(bounty) = bounty {
-						if bounty.value <= *budget_remaining {
-							*budget_remaining -= bounty.value;
-
-							// jump through the funded phase if we're already approved with curator
-							if let BountyStatus::ApprovedWithCurator { curator } = &bounty.status {
-								bounty.status =
-									BountyStatus::CuratorProposed { curator: curator.clone() };
-							} else {
-								bounty.status = BountyStatus::Funded;
-							}
+	fn process_payment_status(
+		source: &T::Beneficiary,
+		beneficiary: &T::Beneficiary,
+		asset_kind: &T::AssetKind,
+		value: BountyBalanceOf<T, I>,
+		payment_status: &mut PaymentState<PaymentIdOf<T, I>>,
+	) -> DispatchResultWithPostInfo {
+		match payment_status {
+			PaymentState::Failed | PaymentState::Pending => {
+				// attempt payment again from failed/pending status
+				let id = T::Paymaster::pay(source, beneficiary, asset_kind.clone(), value)
+					.map_err(|_| Error::<T, I>::FundingError)?;
+				*payment_status = PaymentState::Attempted { id };
+				return Ok(Pays::No.into());
+			},
+			PaymentState::Succeeded => {
+				// should never happen for funding payments,
+				// if payment succeeds in check_status it is advanced
+				// to new bounty status skipping setting `Succeeded`
+				// payment state
+				return Err(Error::<T, I>::UnexpectedStatus.into());
+			},
+			PaymentState::Attempted { .. } => {
+				// payment already in progress, its status needs to be checked
+				return Err(Error::<T, I>::UnexpectedStatus.into());
+			},
+		}
+	}
 
-							// return their deposit.
-							let err_amount = T::Currency::unreserve(&bounty.proposer, bounty.bond);
-							debug_assert!(err_amount.is_zero());
+	fn check_payment_status_for_funding(
+		bounty_status: &mut Option<BountyStatusOf<T, I>>,
+		payment_status: &mut PaymentState<PaymentIdOf<T, I>>,
+		approved_curator: Option<&T::AccountId>,
+	) -> DispatchResultWithPostInfo {
+		match payment_status {
+			PaymentState::Pending => {
+				// user should try processing payment again, not check its status
+				return Err(Error::<T, I>::UnexpectedStatus.into());
+			},
+			PaymentState::Failed => {
+				// payment is known to have failed, this function should revert
+				// not to pass as valid transaction
+				return Err(Error::<T, I>::UnexpectedStatus.into());
+			},
+			PaymentState::Succeeded => {
+				// should never happen in funding payment, we should
+				// have advanced to the next bounty state on success
+				// from ::Attempted payment status
+				return Err(Error::<T, I>::UnexpectedStatus.into());
+			},
```

Comment:
I would try to simplify this. For example these three cases can be handled with one ensure!. I would try to get rid of the `check_payment_status_for_funding` and `check_payment_status_for_payout` abstractions. If needed you can consider introducing different calls for different payout checks. Also consider that we need to benchmark these new calls, the current setup probably requires a worth case benchmark + some specific cases.